### PR TITLE
Reorganisation of `main` branch

### PR DESCRIPTION
**Pull request by storojs72**
The idea of this PR is following: keep in `main` branch our cryptographic building blocks (for every feature) with their unit-tests and keep rest of all developments, related to e2e verification in appropriate branches ([pasta](https://github.com/lurk-lab/solidity-verifier/tree/pasta), [grumpkin](https://github.com/lurk-lab/solidity-verifier/tree/grumpkin), [zeromorph](https://github.com/lurk-lab/solidity-verifier/tree/zeromorph), [gas-optimizing](https://github.com/lurk-lab/solidity-verifier/tree/gas-optimizing), etc.) together with specific related stuff (e2e contract, data loader, proof/public parameters JSONs, integration testing with cloud Anvil node), without attempting to generalise/stabilise the whole codebase in one place. 

Solidity has limited support (no generics, for example) of tools that allow creating unified codebase for scattered functionalities, so attempting to combine all our developments within single codebase seems to be wasting of efforts. Instead it is better to maintain every "feature" in its own branch, until it would be possible to run the whole e2e verification without OutOfGas exception. When it happens, the next step would be deriving the initial version of e2e contract (by combining all features together - which should happen one time), turning it into Yul / assmebly one and start countdown of "stable" solution versions. And the e2e contract can appear in `main` branch. Until that, since `main` is usually considered as a source of "stable" codebase - we can keep fundamental cryptographic building blocks in `main` - since they are more stable - at reference side, in Arecibo, usually new blocks appear (grumpkin arithmetics, zeromorph, etc.) instead of significant changing the existing ones. 

This approach will allow better modularity and isolation of unstable e2e verification contracts and steps libraries as well as avoiding spending time on temporary combining codebase in the "middle" of its evolution.

### CODE REVIEW

File: .github/workflows/end2end.yml
Comment (tchataigner) [inline .github/workflows/end2end.yml:57]: Did you mean to keep it commented? We could have some conditions to make it run when `!test` is commented, but only on `feature/*` branches?

Also, should we have this flow possible on both issues and PRs or only PRs?

**Review by tchataigner — COMMENTED** [review]
Thanks! I have only a discussion to have on the ci workflow. Otherwise the reorg seems to be as you previsouly specified :)

File: .github/workflows/end2end.yml
Comment (storojs72) [reply .github/workflows/end2end.yml:57]: For the `main` branch we don't need integration testing - though it is relevant for other branches. Let's involve @samuelburnham to provide some input on this - as he initially helped with that stuff. 

> We could have some conditions to make it run when `!test` is commented, but only on feature/* branches

That seems to be quite good approach

Comment (samuelburnham) [reply .github/workflows/end2end.yml:57]: I can tackle running on `feature/*` branches, sounds like a great plan.

> Also, should we have this flow possible on both issues and PRs or only PRs?

PRs only I think, if we want to run on branches without an open PR it can be done via [`workflow_dispatch`](https://github.com/lurk-lab/solidity-verifier/actions/workflows/ci.yml) from https://github.com/lurk-lab/solidity-verifier/blob/main/.github/workflows/ci.yml#L8

**Review by huitseeker — CHANGES_REQUESTED** [review]
Workflows and their triggers are different. Workflows live on the main branch of the repo and trigger according to the conditions on their "on:" section. 
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows
Iow, if we want some jobs to not run on e.g. the master branch's pushes, it should be enough to exclude the master branch pushes from their triggers. No need to comment everything.

Please note there is a branches-ignore attribute you can set on most events: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-pull_request-workflow-based-on-the-head-or-base-branch-of-a-pull-request

**Comment by storojs72** [comment]
@huitseeker, executing integration testing via commenting in the PR is quite convenient. With specific branch protection rules, we can manage whether integration test passing is required or not. For example for `main` it is not required before merging (since we have no e2e contract in main). For `pasta` and `grumpkin` branches, it is required, since we have e2e contracts there. For `zeromorph` and `gas-optimised` branches - it is not yet required, but will be required later, when relevant e2e contracts appear.

**Comment by storojs72** [comment]
I think, we can follow this simple policy for now (using branch protection rules): 
- require unit tests passing for `main` and all derived branches (as they don't have e2e contracts);
- require unit tests and integration tests passing for `pasta` and `grumpkin` branches (as they have e2e contracts). In order to invoke executing integration tests, one of core contributors (from our organisation) should comment `!test` (also it should work for PRs from external contributors).

When we finish composing the e2e contracts for `zeromorph` and `gas-optimizing` or whatever feature in future, we add new branch protection rule for correspondent branch and enjoy.

**Comment by huitseeker** [comment]
@storojs72 I'm not criticizing the merge policy or the tests you want to run on any particular branch. I trust you on that part.
I'm criticizing the specific manner in which you run those tests.

There are two parts to each CI workflow, which correspond to actual distinct areas a workflow file:
1. describing the steps of a particular job (that's the `jobs:` section in GH actions),
2. describing the events that lead to those steps being run (that's the `on:` section in GH actions),

In order to change *when* tests are run, the natural and efficient way to do things is to modify part 2.
If order to remove tests *that will no longer ever be needed on any branch at all*, it makes sense to remove workflow steps from part 1.

I do not understand the logic that leads to *commenting*  steps in part 1, I'd love an explanation.

**Comment by storojs72** [comment]
@huitseeker, I finally got that. Yes, commenting steps of integration testing was not a good idea. In  https://github.com/lurk-lab/solidity-verifier/pull/42/commits/52b17a3a7e0fa0c6dd5a7ca1c01209b7a228aa4c it is actually reverted and only one thing is added - skipping the `integration-tests-e2e` job execution for `main` branch (even if it is somehow triggered)

**Review by huitseeker — APPROVED** [review]
This now makes sense to me, thanks!

File: .github/workflows/end2end.yml
Comment (samuelburnham) [inline .github/workflows/end2end.yml:24]: Have you tested this works? I'm worried this will never return `true`, because `issue_comment` actions are always triggered from the default branch and we haven't checked out the base branch of the PR at this point.

Comment (tchataigner) [reply .github/workflows/end2end.yml:24]: I'm also afraid of this scenario.

Another way of fetching informations on concerned branch might be to use [`github.event.issue.pull_request`](https://docs.github.com/en/webhooks/webhook-events-and-payloads#issue_comment) to get info on the source branch.

**Review by tchataigner — CHANGES_REQUESTED** [review]
I think the e2e CI will not work with the newly added condition.

File: .github/workflows/end2end.yml
Comment (samuelburnham) [reply .github/workflows/end2end.yml:24]: @tchataigner I've managed to find a workaround in https://github.com/lurk-lab/solidity-verifier/pull/45. A nifty alternative is to run `gh pr view <pr-number> --json baseRefName --jq '.baseRefName'`, but GitHub certainly doesn't make this easy to figure out :smile:

Comment (storojs72) [reply .github/workflows/end2end.yml:24]: You are both right. We have https://github.com/lurk-lab/solidity-verifier/pull/45 for tackling with this, and probably we will need merge it first

Comment (storojs72) [reply .github/workflows/end2end.yml]: https://github.com/lurk-lab/solidity-verifier/pull/45 has been merged and it looks like we have this issue resolved

**Review by tchataigner — APPROVED** [review]

**Pull request merged**
